pkg/health/health.go

```diff
@@ -0,0 +1,158 @@
+package health
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/GoogleCloudPlatform/cloud-run-release-operator/internal/metrics"
+	"github.com/GoogleCloudPlatform/cloud-run-release-operator/pkg/config"
+	"github.com/pkg/errors"
+)
+
+// Diagnosis is the information about the health of the revision.
+type Diagnosis struct {
+	EnoughRequests     bool
+	CheckResults       []*CheckResult
+	FailedCheckResults []*CheckResult
+	IsHealthy          bool
+}
+
+// CheckResult is information about a metrics criteria check.
+type CheckResult struct {
+	MetricsType   config.MetricsCheck
+	MinOrMaxValue interface{}
+	ActualValue   interface{}
+	IsCriteriaMet bool
+	Reason        string
+}
+
+// Diagnose attempts to determine the health of a revision.
+//
+// If the minimum number of requests is not met, then health cannot be
+// determined and Diagnosis.EnoughRequests is set to false.
+//
+// Otherwise, all metrics criteria are checked to determine if the revision is
+// healthy.
+func Diagnose(ctx context.Context, provider metrics.Metrics, query metrics.Query,
+	offset time.Duration, minRequests int64, healthCriteria []config.Metric) (*Diagnosis, error) {
+
+	metricsValues, err := CollectMetrics(ctx, provider, query, offset, healthCriteria)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not collect metrics")
+	}
+
+	isHealthy := true
+	var results, failedResults []*CheckResult
+	for i, criteria := range healthCriteria {
+		result := determineResult(criteria.Type, criteria.Min, criteria.Max, metricsValues[i])
+		results = append(results, result)
+
+		if !result.IsCriteriaMet {
+			isHealthy = false
+			failedResults = append(failedResults, result)
+		}
+	}
+
+	return &Diagnosis{
+		EnoughRequests:     true,
+		CheckResults:       results,
+		FailedCheckResults: failedResults,
+		IsHealthy:          isHealthy,
+	}, nil
+}
+
+// CollectMetrics returns an array of values collected for each of the specified
+// metrics criteria.
+func CollectMetrics(ctx context.Context, provider metrics.Metrics, query metrics.Query, offset time.Duration, healthCriteria []config.Metric) ([]interface{}, error) {
+
+	var values []interface{}
+	for _, criteria := range healthCriteria {
+		var value interface{}
+		var err error
+
+		switch criteria.Type {
+		case config.LatencyMetricsCheck:
+			value, err = latency(ctx, provider, query, offset, criteria.Percentile)
+			break
+		case config.ErrorRateMetricsCheck:
+			value, err = errorRate(ctx, provider, query, offset)
```

Comment:
add logs

Reply:
This will be done in a following PR after we integrate this package with the `rollout`